test/integration/retry_budget_integration_test.cc

```diff
@@ -0,0 +1,284 @@
+    upstream_request_->encodeHeaders(error_response_headers_, !include_response_body);
+    if (include_response_body) {
+      upstream_request_->encodeData("response body", true);
+    }
+    bool expect_conn_closed = include_response_body && upstreamProtocol() == Http::CodecType::HTTP1;
+    markUpstreamRequestComplete(expect_conn_closed);
+    codec_clients_.push_back(std::move(codec_client));
+
+    if (expect_response) {
+      ASSERT_TRUE(response->waitForEndStream());
+    } else {
+      unfinished_responses_.push_back(std::move(response));
+    }
+  }
+
+  void TearDown() override {
+    for (auto& upstream_request : unfinished_upstream_requests_) {
+      upstream_request->encodeHeaders(default_response_headers_, true);
+      ASSERT_TRUE(upstream_request->complete());
+    }
+    for (auto& response : active_request_responses_) {
+      ASSERT_TRUE(response->waitForEndStream());
+    }
+    for (auto& conn : busy_upstream_conns_) {
+      if (conn != nullptr) {
+        AssertionResult result = conn->close();
+        RELEASE_ASSERT(result, result.message());
+        result = conn->waitForDisconnect();
+        RELEASE_ASSERT(result, result.message());
+        conn.reset();
+      }
+    }
+    for (auto& conn : unusable_upstream_conns_) {
+      if (conn != nullptr) {
+        AssertionResult result = conn->close();
+        RELEASE_ASSERT(result, result.message());
+        result = conn->waitForDisconnect();
+        RELEASE_ASSERT(result, result.message());
+        conn.reset();
+      }
+    }
+    for (auto& client : codec_clients_) {
+      if (client->rawConnection().state() == Envoy::Network::Connection::State::Open) {
+        // Manually abort ongoing requests (retries still in backoff)
+        client->rawConnection().close(Envoy::Network::ConnectionCloseType::Abort);
+      }
+      client->close();
+    }
+    unfinished_upstream_requests_.clear();
+    active_request_responses_.clear();
+    busy_upstream_conns_.clear();
+    codec_clients_.clear();
+  }
+
+protected:
+  std::vector<Envoy::IntegrationCodecClientPtr> codec_clients_;
+  std::vector<Envoy::IntegrationStreamDecoderPtr> unfinished_responses_;
+  std::vector<Envoy::IntegrationStreamDecoderPtr> active_request_responses_;
+  std::vector<Envoy::FakeStreamPtr> unfinished_upstream_requests_;
+  Http::TestResponseHeaderMapImpl error_response_headers_{{":status", "500"}};
+
+  std::vector<Envoy::FakeHttpConnectionPtr> busy_upstream_conns_;
+  Envoy::FakeHttpConnectionPtr free_upstream_conn_;
+  std::vector<Envoy::FakeHttpConnectionPtr> unusable_upstream_conns_;
+};
+
+INSTANTIATE_TEST_SUITE_P(Protocols, RetryBudgetIntegrationTest,
+                         testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams(
+                             {Http::CodecType::HTTP1, Http::CodecType::HTTP2},
+                             {Http::CodecType::HTTP1, Http::CodecType::HTTP2,
+                              Http::CodecType::HTTP3})),
```

Comment:
Including h3 in the downstream codec causes it to segfault and I'm not sure why. I don't think it has anything to do with my changes since they should only impact the upstream if anything. I'm pretty sure this is either a bug in the integration testing setup, or maybe something I'm doing wrong in my connection management code in the test.